.circleci/config.yml

```diff
@@ -38,80 +42,128 @@ jobs:
-            # 1) every time a commit is pushed to `master`, a Snap build for `edge` channel
-            #    is triggered automatically via GitHub webhook on https://build.snapcraft.io/user/VirtusLab/git-machete,
-            # 2) every time a https://github.com/r-ryantm bot runs on NixOS/nixpkgs repository,
-            #    Nix git-machete package is updated with the fresh version/package hash according to the data from PyPI.
+
+  deploy_from_develop:
+    executor: ubuntu_executor
+    steps:
+      - deploy:
+          name: '[Dry run] Deploy to PyPI'
+          # Side effects: does an actual release, but to the test PyPI
+          command: bash ci/pypi/ci-deploy.sh --dry-run
+
+      - deploy:
+          name: '[Dry run] Update Homebrew tap'
+          # Prerequisites: artifacts must be released to test PyPI first
+          command: bash ci/homebrew-tap/ci-deploy.sh --dry-run
+
+      - deploy:
+          name: '[Dry run] Build & upload deb package to PPA'
+          # Side effects: potentially pushes a regenerated Docker image
+          command: bash ci/deb-ppa-upload/ci-deploy.sh --dry-run
+
+      - deploy:
+          name: '[Dry run] Build rpm package'
+          # Side effects: potentially pushes a regenerated Docker image
+          # Other than that, it just builds the package without publishing anything; there's no special dry-run mode
+          command: bash ci/rpm/ci-run.sh
+
+  deploy_from_master:
+    executor: ubuntu_executor
+    steps:
+      - deploy:
+          name: 'Deploy to PyPI'
+          command: bash ci/pypi/ci-deploy.sh
+
+      - deploy:
+          name: 'Update Homebrew tap'
+          # Prerequisites: artifacts must be released to PyPI first
+          command: bash ci/homebrew-tap/ci-deploy.sh
+
+      - deploy:
+          name: 'Build & upload deb package to PPA'
+          command: bash ci/deb-ppa-upload/ci-deploy.sh
+
+      - deploy:
+          name: 'Build rpm package'
+          command: bash ci/rpm/ci-run.sh
+
+      # Let's do this at the very end to avoid opening a GH release (and notifying the watchers)
+      # in case anything goes wrong in the previous steps.
+      - deploy:
+          name: 'Create GitHub release'
+          # Prerequisites: RPM package must be in dist/
+          command: |
+            sudo apt-get update
+            sudo apt-get install hub
+            bash ci/github-releases/ci-deploy.sh
+
+      # Note: there are also releases happening outside of our CI:
+      # 1) every time a commit is pushed to `master`, a Snap build for `edge` channel
+      #    is triggered automatically via GitHub webhook on https://build.snapcraft.io/user/VirtusLab/git-machete,
+      # 2) every time a https://github.com/r-ryantm bot runs on NixOS/nixpkgs repository,
+      #    Nix git-machete package is updated with the fresh version/package hash according to the data from PyPI.
+
+workflows:
+  version: 2
+  test-deploy:
+    jobs:
+      - initial_checks
```

Comment:
Just a loose suggestion, might make sense or not... maybe create a no-op/empty job with a super-lightweight executor that would depend on all of `initial_checks...tests_py39`? Then it can be used as both the required check for Github and as `requires` for `deploy_from_develop`/`deploy_from_master`

Reply:
It will look better in the code, but it will skew the circle ci diagram (in reality deployment runs are waiting for tests, not for empty step) and it will drop the performance a bit, because CI still have to assign a VM to it, which sometimes takes a moment.